src/extend.cpp

```diff
@@ -127,48 +133,143 @@ sparse_coo extend(const sparse_coo &sparse_rr_mat,
   std::vector<float> dists_all = combine_vectors(dists, len);
   std::vector<long> i_vec_all = combine_vectors(i_vec, len);
   std::vector<long> j_vec_all = combine_vectors(j_vec, len);
-
   return (std::make_tuple(i_vec_all, j_vec_all, dists_all));
 }
 
-sparse_coo lower_rank(const sparse_coo &sparse_rr_mat, const size_t n_samples,
-                      const size_t kNN) {
+sparse_coo lower_rank(const sparse_coo &sparse_rr_mat,
+                      const size_t n_samples,
+                      const size_t kNN,
+                      bool reciprocal_only,
+                      bool count_unique_distances,
+                      size_t num_threads = 1) {
+  // Data structures for iteration
+  size_t len = 0;
   std::vector<long> row_start_idx = row_start_indices(sparse_rr_mat, n_samples);
-
-  // ijv vectors
-  std::vector<float> dists;
-  std::vector<long> i_vec;
-  std::vector<long> j_vec;
   std::vector<float> dist_vec = std::get<2>(sparse_rr_mat);
   const std::vector<long> j_sparse = std::get<1>(sparse_rr_mat);
+  // ijv vectors
+  std::vector<std::vector<float>> dists(n_samples);
+  std::vector<std::vector<long>> i_vec(n_samples);
+  std::vector<std::vector<long>> j_vec(n_samples);
+#pragma omp parallel for schedule(static) num_threads(num_threads) reduction(+:len)
```

Comment:
If this taks some time you may want to add an interrupt handler here too, but not worth it if quick

Reply:
Will see how slow this gets with longer profiling!